python/ray/data/datasource/databricks_uc_datasource.py

```diff
@@ -0,0 +1,188 @@
+from urllib.parse import urlencode, urljoin
+
+import pyarrow
+import requests
+
+from ray.data.block import BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask
+from ray.util.annotations import PublicAPI
+
+logger = logging.getLogger(__name__)
+
+
+class _DatabricksUCReader(Reader):
+    def __init__(
+        self,
+        host: str,
+        token: str,
+        warehouse_id: str,
+        catalog: str,
+        schema: str,
+        query: str,
+    ):
+        self.host = host
+        self.token = token
+        self.warehouse_id = warehouse_id
+        self.catalog = catalog
+        self.schema = schema
+        self.query = query
+
+        url_base = f"https://{self.host}/api/2.0/sql/statements/"
+
+        payload = json.dumps(
+            {
+                "statement": self.query,
+                "warehouse_id": self.warehouse_id,
+                "wait_timeout": "0s",
+                "disposition": "EXTERNAL_LINKS",
+                "format": "ARROW_STREAM",
+                "catalog": self.catalog,
+                "schema": self.schema,
+            }
+        )
+
+        req_auth = ("token", self.token)
+        req_headers = {"Content-Type": "application/json"}
+
+        response = requests.post(
+            url_base,
+            auth=("token", self.token),
+            headers={"Content-Type": "application/json"},
+            data=payload,
+        )
+        response.raise_for_status()
+        statement_id = response.json()["statement_id"]
+
+        state = response.json()["status"]["state"]
+
+        logger.info(f"Waiting for query {query} exeuction result.")
+        try:
+            while state in ["PENDING", "RUNNING"]:
+                time.sleep(1)
+                response = requests.get(
+                    urljoin(url_base, statement_id) + "/",
+                    auth=req_auth,
+                    headers=req_headers,
+                )
+                response.raise_for_status()
+                state = response.json()["status"]["state"]
+        except KeyboardInterrupt:
+            # User cancel the command, so we cancel query execution.
+            requests.post(
+                urljoin(url_base, f"{statement_id}/cancel"),
+                auth=req_auth,
+                headers=req_headers,
+            )
+            try:
+                response.raise_for_status()
+            except Exception as e:
+                logger.warning(
+                    f"Canceling query {query} execution failed, reason: {repr(e)}."
+                )
+            raise
+
+        if state != "SUCCEEDED":
+            raise RuntimeError(f"Query {self.query} execution failed.")
+
+        manifest = response.json()["manifest"]
+        is_truncated = manifest["truncated"]
+
+        if is_truncated:
+            logger.warning(
+                f"The result dataset of '{query}' exceeding 100GiB and it is truncated."
+            )
+
+        chunks = manifest["chunks"]
```

Comment:
The manifest contains the first chunk so we could avoid re-reading that, right? Don't think the additional code complexity is worth the performance improvement, but just wanted to make sure that was considered

Reply:
> The manifest contains the first chunk

No, the manifest only contains metadata of every chunk, but it does not contain first chunk data bytes :)

Reply:
Ah, I meant does `manfiest["result"]` contain the external links for the first chunk? In any case, it's probably not worth the complexity of this sort of optimization even if it's possible